### PR TITLE
Add deprecated annotations in fields

### PR DESCRIPTION
**Pull request by viniciusmelquiades**
This PR adds support for the `@deprecated` [annotation](https://spec.graphql.org/June2018/#sec--deprecated) in fields and enum values.

### CODE REVIEW

**Comment by comigor** [comment]
Hey there @viniciusmelquiades! That seems amazing, thanks for you contribution!

We're currently focusing efforts into the [**beta**](https://github.com/comigor/artemis/tree/beta) branch, because a lot of breaking changes can happen/already happened there without many consequences. The idea is to eventually merge it into master. Beta branch is a lot more advanced than master, with a lot of bugfixes.

How do you feel rebasing it there? Some of changes you did (named annotation list on ClassProperty and EnumValue) are already done there, which could help on the process.

Please, reach me in any doubt or if you need any help!

**Comment by comigor** [comment]
BTW, it seems wise to create a CONTRIBUTING file of some kind, to avoid re-works.

**Comment by viniciusmelquiades** [comment]
I can rebase it, no problem. But it will take a while.

And yes, you definitely should create a CONTRIBUTING file.

**Comment by viniciusmelquiades** [comment]
I've converted this to a draft while I work on this change.

**Comment by viniciusmelquiades** [comment]
@comigor instead of rebasing I just copied and adapted the tests to the beta branch and reimplemented the code (this was somehow the easiest part).

One of the cheks is failing, but I have no idea why.

**Comment by vasilich6107** [comment]
I'll take a look

File: lib/generator.dart [hunk 1]
```diff
@@ -342,6 +342,11 @@ Make sure your query is correct and your schema is updated.''');
     annotations.add('JsonKey(name: \'${name.name}\')');
   }
 
+  final fieldDirectives =
+      regularField?.directives ?? regularInputField?.directives;
+
+  _addDeprecatedAnnotationIfNecessary(annotations, fieldDirectives);
```
Comment (vasilich6107) [inline lib/generator.dart:348]: It is not obvious what is happening with annotations inside this function.
`annotations.addAll(_proceedDeprecated(fieldDirectives));` could be more clear approach

File: lib/generator.dart [hunk 2]
```diff
@@ -350,6 +355,28 @@ Make sure your query is correct and your schema is updated.''');
   );
 }
 
+void _addDeprecatedAnnotationIfNecessary(
```
Comment (vasilich6107) [inline lib/generator.dart:358]:
```
List<String> _proceedDeprecated(
  List<DirectiveNode> directives,
) {
  final annotations = <String>[];

  final deprecatedDirective = directives?.firstWhere(
    (directive) => directive.name.value == 'deprecated',
    orElse: () => null,
  );

  if (deprecatedDirective != null) {
    final reasonValueNode = deprecatedDirective?.arguments
        ?.firstWhere((argument) => argument.name.value == 'reason')
        ?.value;

    final reason = reasonValueNode is StringValueNode
        ? reasonValueNode.value
        : 'No longer supported';

    annotations.add("Deprecated('$reason')");
  }

  return annotations;
}
```

File: lib/generator.dart [hunk 3]
```diff
@@ -635,11 +662,20 @@ class _CanonicalVisitor extends RecursiveVisitor {
 
     enums.add(EnumDefinition(
       name: enumName,
-      values: node.values.map((eV) => EnumValue(name: eV.name.value)).toList()
-        ..add(ARTEMIS_UNKNOWN),
+      values: node.values.map(_getEnumValue).toList()..add(ARTEMIS_UNKNOWN),
```
Comment (vasilich6107) [inline lib/generator.dart:665]: I do not think that dedicated function is required
```
enums.add(EnumDefinition(
      name: enumName,
      values: node.values
          .map((eV) => EnumValue(
                name: eV.name.value,
                annotations: _proceedDeprecated(eV.directives),
              ))
          .toList()
            ..add(ARTEMIS_UNKNOWN),
    ));
```

File: lib/generator/data/enum_definition.dart [hunk 4]
```diff
@@ -29,11 +29,18 @@ class EnumDefinition extends Definition with DataPrinter {
 /// Enum value name
 class EnumValue extends Name with DataPrinter {
   /// Instantiate a enum value name definition.
-  EnumValue({@required String name}) : super(name: name);
+  EnumValue({
```
Comment (vasilich6107) [inline lib/generator/data/enum_definition.dart:32]: I think it is better to move towards the same approach as in classes.

We will get 
`EnumDefinition` -> for enum
`EnumValueDefinition` -> for name and annotations
and rename `EnumValue` => `EnumValueName`

**Comment by vasilich6107** [comment]
@viniciusmelquiades
Hi. Will you have time to fix this?
If not - I can take a valuable part of this PR implement this feature with proposed changes

**Comment by viniciusmelquiades** [comment]
@vasilich6107 Yes, I will. It got lost in my notifications but I'll take a look at it.

**Comment by viniciusmelquiades** [comment]
@vasilich6107 I made the changes you requested and rebased the code on `beta`

**Comment by vasilich6107** [comment]
Thanks i'll check

**Comment by vasilich6107** [comment]
Looks ok)
@comigor could you look why the CI tests failed?

**Comment by comigor** [comment]
It seems that GitHub Actions bot can't comment in the PR unless the branch is from one of the maintainers, which doesn't make sense IMO. I'll try to circumvent that.

**Comment by comigor** [comment]
I'll test some actions stuff in this PR

**Comment by comigor** [comment]
GitHub [forcibly redacts encrypted secrets to be used on forks](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#using-encrypted-secrets-in-a-workflow:~:text=secrets%20are%20not%20passed%20to%20the%20runner%20when%20a%20workflow%20is%20triggered%20from%20a%20forked%20repository) :(

**Comment by comigor** [comment]
And it makes sense: people would be able to fork the repo, change the actions, retrieve the secret and do terrible stuff

**Comment by comigor** [comment]
So there's no fix in this, and the person who's creating the PR will have to check CI logs to see what's wrong.

In this case it's both:
```
You must bump the version on pubspec!
```

```
Analyzer has failed!
Analyzing workspace...
  hint • Unused import: 'package:artemis/generator/helpers.dart'. • lib/generator/data/enum_value_definition.dart:3:8 • unused_import
1 hint found.
```

**Pull request merged**